internal/selfupdate/download.go

```diff
@@ -74,33 +66,50 @@ func extractToFile(buf []byte, filename, target string, printf func(string, ...i
 		rd = file
 	}
 
-	err = os.Remove(target)
-	if os.IsNotExist(err) {
-		err = nil
-	}
+	// Write everything to a temp file
+	dir := filepath.Dir(target)
+	new, err := os.CreateTemp(dir, "restic")
 	if err != nil {
-		return fmt.Errorf("unable to remove target file: %v", err)
+		return err
 	}
 
-	dest, err := os.OpenFile(target, os.O_CREATE|os.O_EXCL|os.O_WRONLY, mode)
+	n, err := io.Copy(new, rd)
 	if err != nil {
+		new.Close()
+		os.Remove(new.Name())
 		return err
 	}
-
-	n, err := io.Copy(dest, rd)
-	if err != nil {
-		_ = dest.Close()
-		_ = os.Remove(dest.Name())
+	if err = new.Sync(); err != nil {
+		return err
+	}
+	if err = new.Close(); err != nil {
 		return err
 	}
 
-	err = dest.Close()
-	if err != nil {
+	printf("saved %d bytes in %v\n", n, new.Name())
+
+	mode := os.FileMode(0755)
+	// attempt to find the original mode
+	if fi, err := os.Lstat(target); err == nil {
+		mode = fi.Mode()
+	}
+
+	// Rename (rather than remove) the running version. The running binary will be locked
+	// on some platforms (Windows) and cannot be removed while executing.
+	backup := filepath.Join(dir, filepath.Base(target)+".bak")
```

Comment:
That will also leave behind a backup file on platforms where the running binary can be removed. What using the backup file codepath only on windows and just move the new version to the target everywhere else?